### ISSUE TITLE
Installation process saves binary in wrong location

### ISSUE THREAD

**Issue by dgg** [nodkz/mongodb-memory-server#168]
In version 5.0.1, when the package is installed the binary is not downloaded to the default location (_./node_modules/.cache/mongodb-memory-server/mongodb-binaries_) but to _./node_modules/mongodb-memory-server/node_modules/.cache/mongodb-memory-server/mongodb-binaries_ instead.
When used for the first time then, it will download the binary to the correct location (and the test will probably time out).

Oddly enough, if skipping installation with `MONGOMS_DISABLE_POSTINSTALL=1` and executing `node ./node_modules/mongodb-memory-server/postinstall.js` afterwards (after enabling the installation, that is), the binary is downloaded to the right location.

Here is the installation output when run in Windows:

```bash
{project_root}$ npm install mongodb-memory-server

> mongodb-memory-server@5.0.1 postinstall {project_root}\node_modules\mongodb-memory-server
{project}> node ./postinstall.js

mongodb-memory-server: checking MongoDB binaries cache...
mongodb-memory-server: binary path is {project_root}\node_modules\mongodb-memory-server\node_modules\.cache\mongodb-memory-server\mongodb-binaries\4.0.5\mongod.exe
+ mongodb-memory-server@5.0.1
added 4 packages from 3 contributors in 51.462s
```

Ideally, setting the version in either environment (or _package.json_ in the future) will download the binary in the right location at install time.

**Comment by nodkz** [comment]
Please try to 5.0.2.

Will be published from minute to minute.

**Comment by dgg** [comment]
Uhm, it did not seem to do the trick:

```bash
{project_root}> npm install mongodb-memory-server

> mongodb-memory-server@5.0.2 postinstall {project_root}\node_modules\mongodb-memory-server
> node ./postinstall.js

mongodb-memory-server: checking MongoDB binaries cache...
mongodb-memory-server: binary path is {project_root}\node_modules\mongodb-memory-server\node_modules\.cache\mongodb-memory-server\mongodb-binaries\4.0.5\mongod.exe
+ mongodb-memory-server@5.0.2
added 4 packages from 3 contributors in 46.254s
```

I am getting the same output and the binary is downloaded to the wrong location.

**Comment by nodkz** [comment]
Very strange. It works for me as expected on an absolutely fresh install.
Do you use it under Win? Maybe somewhere problem with backslashes...

Download starts here https://github.com/nodkz/mongodb-memory-server/blob/master/packages/mongodb-memory-server-core/src/util/MongoBinary.ts#L104

If you are able to debug more closely, it will be very helpful.
Thanks.

**Comment by dgg** [comment]
I do not have a non-windows machine with me at the moment, but I will look at it this evening.
I will also try to debug the existing code in my windows machine and let you know.

I 'll get back to you

Thanks for your help

**Comment by dgg** [comment]
With a fresh installation, in a Mac computer, it works just fine:

```bash
{clean_project}$ npm install

> mongodb-memory-server@5.0.2 postinstall {clean_project}/node_modules/mongodb-memory-server
> node ./postinstall.js

mongodb-memory-server: checking MongoDB binaries cache...
mongodb-memory-server: binary path is {clean_project}/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.5/mongod
```

I will be checking with the same project I was checking this morning in the Mac shortly...

**Comment by dgg** [comment]
With the same project as I was checking this morning, but in a Mac computer, it works just fine.

```
$ nvm --version
0.34.0
$ nvm which current
~/.nvm/versions/node/v10.15.0/bin/node
$ node --version
v10.15.0
$ npm --version
6.9.0
```
I'll check in Linux (WSL) and debug further in Windows shortly.

**Comment by dgg** [comment]
Installation also works as it should in Linux (WSL).

It **does not work** with a clean project in Windows (neither using "classic" command prompt or Powershell)
```shell
$ npm install mongodb-memory-server --save-dev                                                                    
                                                                                                                  
> mongodb-memory-server@5.0.2 postinstall {clean_project}\node_modules\mongodb-memory-server                          
> node ./postinstall.js                                                                                           
                                                                                                                  
mongodb-memory-server: checking MongoDB binaries cache...                                                         
mongodb-memory-server: binary path is {clean_project}\node_modules\mongodb-memory-server\node_modules\.cache\mongodb-memory-server\mongodb-binaries\4.0.3\mongod.exe                                                                    
```

I'll look more in depth what is going on in the file you pointed out. 😞

**Comment by nodkz** [comment]
I think that problem in regexp or maybe in path.sep value which makes regexp invalid. It breaks while clause. 
https://github.com/nodkz/mongodb-memory-server/blob/master/packages/mongodb-memory-server-core/src/util/MongoBinary.ts#L109

**Comment by dgg** [comment]
@nodkz I think I have found the culprit. As you suspected it has to do with path separators.

Since a regex is used to test the path we are execution the script in, (https://github.com/nodkz/mongodb-memory-server/blob/master/packages/mongodb-memory-server-core/src/util/MongoBinary.ts#L109).

In Linux-like file systems, where the separator is a forward slash `> path.sep --> '/'`, the pattern to test against is _node_modules/mongodb-memory-server_ and the test evaluates to `true`, making the script to "climb up" the hierarchy to locate the cache dir.

In Windows, where the separator is backslash `> path.sep --> '\\'`, the pattern to test against is _node_modules\\mongodb-memory-server_ and the test evaluates to `false`, making the script to **not** "climb up" the hierarchy and placing the cache dir inside the package installation.

I am no reg ex guru but I think that back slashes are used to escape characters and that is why the pattern evaluates to false.

A somewhat dirty fix would be replacing single backslashes by double backslashes:
```
> new RegExp(`node_modules${path.sep}mongodb-memory-server`).test(nodeModulesDLDir)                      
false                                                                                                    
> new RegExp(`node_modules${path.sep}mongodb-memory-server`.replace('\\', '\\\\')).test(nodeModulesDLDir)
true                                                                                                     
```
I am not sure if there is a better way to fix it, though, but the expression evaluates to `true` both in Windows and Linux.

**Comment by hmarlo** [comment]
I'm experiencing the same on Win 10 Home. 

Autodownload nests the binaries inside the `mongo-memory-server/node_modules` folder, while downloading using `node .\node_modules\mongodb-memory-server\postinstall.js` do it correctly.

If I move to the folder instead (`cd .\node_modules\mongodb-memory-server\`) and then execute `postinstall.js`, I get the same wrong download dir that autodownload do. Then it seems that the working directory is the problem here.

**Comment by dir01** [comment]
I created repository reproducing the issue https://gist.github.com/dir01/ea608ea7d804066535a8e2ffb725bf31

`npm install mongodb-memory-server@5.1.1` downloads binaries to `node_modules/mongodb-memory-server/mongodb-binaries`, but then upon the first start, the library downloads binaries to `node_modules/.cache`

**Comment by dahlbyk** [comment]
What @dir01 reports seems to be a separate issue, but I can confirm what @dgg reports. If I'm understanding the `../..` logic correctly, we can equivalent check `endsWith()` and avoid the RegExp escape issue: https://github.com/nodkz/mongodb-memory-server/pull/194.

**Comment by dahlbyk** [comment]
Confirmed fixed in v5.1.4. Thanks for quick release, @nodkz!

![image](https://user-images.githubusercontent.com/133987/59310353-b7875600-8c6b-11e9-9a0f-84176d9f722b.png)
